### ISSUE TITLE
Relate view with Collection, where view has different _ids than the source collection?

### ISSUE THREAD

**Issue by janus-reith** [rwynn/monstache#351]
I have a collection `Catalog` and a view `view_CatalogGroups`.
`view_CatalogGroups` unwinds from `Catalog`, and optionally appends text to the _id field.
So in my case, I have around ~1000 documents in my `Catalog`, resulting in 1.500 documents in my view `view_CatalogGroups`.
Therefore the view `view_CatalogGroups` has some new/altered _ids, but also has the same _id in same cases. The other way around, Catalog has some _ids which are not present in their unaltered form in `view_CatalogGroups` anymore.

Now while I can't use the _id field anymore to properly reindex the view, I have a field `product._id`.
So, one changed document in the `Catalog` might trigger changes for three items in `view_CatalogGroups` where all have the same `product._id`
`product` is just an object, not an array, so I exptect `product._id` to cause no issues.

My problem now is: The Resulting configuration I get would either not reflect changes from `Catalog`, or also include _id objects from `Catalog` which are not present in the`view_CatalogGroups`.
E.g. one doc in `Catalog` is split into 3 docs in `view_CatalogGroups` that all have a seprate _id now (which is not always the case). The result in the Index will be 4 _ids.

My configuration currently is:

```
direct-read-namespaces = ["mydb.Catalog", "mydb.view_CatalogGroups"]
change-stream-namespaces = ["mydb.Catalog"]

# CatalogGroups view

# This mapping is necessary for updates from Catalog, but creates additional entries present in Catalog but not in the view
[[mapping]]
namespace = "mydb.Catalog"
index = "CatalogGroups"

# Is this even necessary?
[[mapping]]
namespace = "mydb.view_CatalogGroups"
index = "CatalogGroups"

# I expected this to make sure that not the _id in Catalog is used, but product.productId resulting in the correct documents. keep-src = false also had no effect on the issue
[[relate]]
namespace = "mydb.Catalog"
with-namespace = "mydb.view_CatalogGroups"
src-field = "product.productId"
match-field = "product.productId"
keep-src = true
```

With that, I get the aforementioned _ids from `Catalog` which should not be present.
Would realy appreciate your help! :)

**Comment by janus-reith** [comment]
I noticed that in https://github.com/rwynn/monstache/issues/323 the dot is encoded.
I therefore tried to alter my configuration accordingly:
```
src-field = "product\uff0eproductId"
match-field = "product\uff0eproductId"
```

This however doesn't seem to have any effect.

**Comment by janus-reith** [comment]
I noticed that it works with only one mapping to the view and one relate if I use a toplevel field for src-field/match-field. Is a nested field like `product.productId` not supported?

**Comment by rwynn** [comment]
@janus-reith I think this may be similiar to https://github.com/rwynn/monstache/issues/347. 

The new `dot-notation` option on a `relate` is committed now, but not yet into a release.  That should happen soon.  In the meantime, if you are able to build from e.g. `rel6` branch, then you could set 
`dot-notation`to `true`.

**Comment by rwynn** [comment]
My understanding is that Issue #347 is not just applicable to arrays. I believe that there is a difference in MongoDB between a query like `{"foo.bar": 1}` and `{foo: {bar: 1}}`.  The former matches regardless of the existence of additional fields in `foo`, while the latter requires `foo` to contain only 1 field `bar`.  

```
rs1:PRIMARY> db.a.insert({foo: {bar: 1, baz:2}})
rs1:PRIMARY> db.a.find({foo: {bar: 1}});
rs1:PRIMARY> db.a.find({"foo.bar": 1});
{ "_id" : ObjectId("5e6a3193556f3e9e708e5122"), "foo" : { "bar" : 1, "baz" : 2 } }
```
I initially made the mistake of defaulting to have the query formed like `{foo: {bar: 1}}` when nested fields are detected.   This is a valid query but not usually what is desired since it requires an exact match.  

To correct this going forward a `dot-notation` option is added that when `true` will change the query to be like `{"foo.bar": 1}`

**Comment by janus-reith** [comment]
- @rwynn Yes, thanks for the quick answer, that sounds reasonable and you' re correct, syntax-wise I could query both an array or object in mongodb this way so this is most likely the case.

Im just wondering why `dot-notation` would not be the default, I guess this is for backwards-compatibility?
Because as fields containing a dot like discussed in some other issues also wouldn't work this way, it seems like the only actual usecase for an exact match like you described here really is uncommon.

**Comment by janus-reith** [comment]
BTW I just noticed there already is a new release, amazing work!

**Comment by rwynn** [comment]
@janus-reith yes, `dot-notation` eq `true` is not the default for backwards compatibility.  I try to make sure there are no surprises and that configurations will not produce different results on upgrade if possible.
